File: Langauge/set/compiler/java/src/main/java/set/antlr4/Function.java
package set.antlr4;

import java.util.List;
import java.util.Map;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import set.antlr4.SETParser.ExpressionContext;

public class Function {
	private String id;
    private List<TerminalNode> params;
    private ParseTree block;

    public Function(String id, List<TerminalNode> params, ParseTree block) {
    	this.id = id;
        this.params = params;
        this.block = block;
    }
    
    public SETValue invoke(List<ExpressionContext> params, Map<String, Function> functions, Scope scope) {
        if (params.size() != this.params.size()) {
            throw new RuntimeException("Illegal Function call");
        }
        scope = new Scope(scope); // create function scope
        EvalVisitor evalVisitor = new EvalVisitor(scope, functions); 
        for (int i = 0; i < this.params.size(); i++) {
            SETValue value = evalVisitor.visit(params.get(i));
            scope.assignParam(this.params.get(i).getText(), value);
        }
        SETValue ret = SETValue.VOID;
        try {
        	evalVisitor.visit(this.block);
        } catch (ReturnValue returnValue) {
        	ret = returnValue.value;
        }
        return ret;
    }
}

File: Music/Langauge/set/compiler/cpp/main/cpp/set/antlr4/Function.cpp
#include "Function.h"
#include "Scope.h"
#include "TLValue.h"
#include "EvalVisitor.h"
#include "ReturnValue.h"

namespace tl
{
	namespace antlr4
	{
		using org::antlr::v4::runtime::tree::ParseTree;
		using org::antlr::v4::runtime::tree::TerminalNode;
		using tl::antlr4::TLParser::ExpressionContext;

		Function::Function(const std::wstring &id, std::vector<TerminalNode*> &params, ParseTree *block)
		{
			this->id = id;
			this->params = params;
			this->block = block;
		}

		TLValue *Function::invoke(std::vector<ExpressionContext*> &params, std::unordered_map<std::wstring, Function*> &functions, Scope *scope)
		{
			if (params.size() != this->params.size())
			{
				std::exception tempVar(L"Illegal Function call");
				throw &tempVar;
			}
			scope = new Scope(scope); // create function scope
			EvalVisitor *evalVisitor = new EvalVisitor(scope, functions);
			for (int i = 0; i < this->params.size(); i++)
			{
				TLValue *value = evalVisitor->visit(params[i]);
				scope->assignParam(this->params[i]->getText(), value);
			}
			TLValue *ret = TLValue::VOID;
			try
			{
				evalVisitor->visit(this->block);
			}
			catch (ReturnValue returnValue)
			{
				ret = returnValue->value;
			}
			return ret;
		}
	}
}
